File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    bool isValid(string& s,int st,int en,unordered_map<char,int>& dict){  
        for(int i=st;i<=en;i++){
            if(dict[s[i]]>1) return false;
        }

        return true;
    }

    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> dict;
        int anc=0,res=0;
        for(int i=0;i<s.length();i++){
            dict[s[i]]++;
            while(!isValid(s,anc,i,dict)){
                dict[s[anc]]--;
                anc++;
            }
            res=max(res,i-anc+1);
        }

        return res;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def checkDup(self,freq)->bool: 
        for val in freq.values(): 
            if val>1: 
                return True 
        return False 
    def lengthOfLongestSubstring(self, s: str) -> int:
        anc,res = 0,0
        freq = {}
        for idx,val in enumerate(s): 
            freq[val]=freq.get(val,0)+1 
            while anc<idx and self.checkDup(freq): 
                freq[s[anc]]-=1 
                anc+=1 
            res = max([res,idx-anc+1])
        return res